File: html2md/lib/post_processor.py
"""Post-processor — clean up Markdown output."""

import re


def process(md: str) -> str:
    """Clean Markdown: strip trailing spaces, dedup blank lines, trim edges."""
    lines = [line.rstrip() for line in md.splitlines()]
    # collapse 3+ blank lines → 1
    cleaned = re.sub(r"\n{3,}", "\n\n", "\n".join(lines))
    return cleaned.strip()

File: html2md/src/post_processor.rs
/// Clean Markdown output: strip trailing spaces, collapse 3+ blanks → 1, trim edges.
///
/// # Example
///
/// ```
/// use html2md::post_processor::process;
///
/// let out = process("hello\n\n\n\n\nworld");
/// assert_eq!(out, "hello\n\nworld");
/// ```
pub fn process(md: &str) -> String {
    // Strip trailing whitespace from each line
    let lines: Vec<String> = md.lines().map(|l| l.trim_end().to_string()).collect();
    // Collapse 3+ consecutive newlines → 1
    let mut joined = lines.join("\n");
    let result = loop {
        let collapsed = joined.replace("\n\n\n", "\n\n");
        if collapsed == joined {
            break collapsed;
        }
        joined = collapsed;
    };
    result.trim().to_string()
}
